### PR TITLE
Mon 3958 key cloak addon centreon authentication

### PR DESCRIPTION
**Pull request by lpinsivy**
## Description

This feature add to Centreon Web the ability for user to login in Centreon with [Keycloak SSO](https://www.keycloak.org/about.html) provider.

- There are just some additions in centreonAuth.SSO.class.php file to allow Centreon call Keycloak API.
- A new section is added in 'Administration/Parameters' to provide all informations needed for Keycloak API.
- The documentation is up to date in HowTo section just after old SSO login chapter.

Based on PR #6572, rebased on master branch with fixes

Tested on 4.3.0 and 6.0.1

Thanks to @wallforfry

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x (master)

<h2> How this pull request can be tested ? </h2>

From PR #6572 

Create a new Docker Keycloak test instance run :
`$ docker run -it -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin --network host jboss/keycloak`

___
- In your browser go to [http://localhost:8080/auth/admin](http://localhost:8080/auth/admin)
- Login with admin/admin
- In the left panel click to _Clients_
- Click on _Create_ button
- In the _Add Client_ form : 
  - Fill _Client ID_ with : Centreon
  - Fill _Root URL_ with http://YOUR_CENTREON_IP/ (supposed your centreon is running on port 80)
  - Click on _Save_ button
- Now for fast config Enable all switchs from _Standard Flow Enabled_ to _Authorization Enabled_
- Check that _Access Type_ is set to _confidential_
- Go to _Installation_ tab
  - Click on the dropdown menu and select _Keycloak OIDC JSON_ and keep this browser tab for later.

___
- Open your Centreon admin panel
- Fill all fields in _Authentication by Keycloak_ with :
  - _Enable Keycloak  authentication_ : [X]
  - _Keycloak Server Url_ : http://YOUR_KEYCLOAK_HOST_IP:8080/auth
  - _Keycloak Redirect Url_ : http://YOUR_CENTREON_IP/centreon/index.php
  - _Keycloak Client Realm_ : **master**
  - _Keycloak Client ID_ : **Centreon**
  - _Keycloak Client Secret_ : **secret provide in the json file in last tab** 

___
- Logout from Centreon
- go to [http://YOUR_CENTREON_URL/centreon/index.php]()
- Click on **Connect by Keycloak** button
- You are redirected to the keycloak login page
- Connect on it with admin/admin
- You are now connect in Centreon with admin account
___

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

**Comment by lpinsivy** [comment]
@wallforfry could you give me the dockerfile to build automatic acceptance test based on your docker?

File: doc/en/administration_guide/keycloak.rst [hunk 1]
```diff
@@ -0,0 +1,21 @@
+.. _keycloak:
+
+**********************
+Implement Keycloak SSO
+**********************
+
+How Keycloak SSO works with Centreon?
```
Comment (sc979) [inline doc/en/administration_guide/keycloak.rst]: line 10 -> no description on "how to". Please change the title

File: doc/fr/administration_guide/keycloak.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
+.. _keycloak:
+
+***********************************
+Implémentation du SSO avec Keycloak
+***********************************
+
+Comment fonctionne le SSO par Keycloak avec Centreon ?
```
Comment (sc979) [inline doc/fr/administration_guide/keycloak.rst]: line 10 -> no description on "how to". Please change the title

File: www/class/centreonAuth.SSO.class.php [hunk 3]
```diff
@@ -55,16 +55,18 @@ public function __construct(
 
         $this->options_sso = $generalOptions;
 
-        if (isset($this->options_sso['sso_enable']) && $this->options_sso['sso_enable'] == 1 &&
-            isset($this->options_sso['sso_header_username']) && $this->options_sso['sso_header_username'] != '' &&
-            isset($_SERVER[$this->options_sso['sso_header_username']])
+        if (isset($this->options_sso['sso_enable'])
+            && $this->options_sso['sso_enable'] == 1
+            && isset($this->options_sso['sso_header_username'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && !empty($this->options_sso['sso_header_username'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 4]
```diff
@@ -55,16 +55,18 @@ public function __construct(
...
+            && $this->options_sso['sso_header_username'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 5]
```diff
@@ -55,16 +55,18 @@ public function __construct(
...
+            && isset($_SERVER[$this->options_sso['sso_header_username']])
         ) {
             $this->sso_username = $_SERVER[$this->options_sso['sso_header_username']];
             if ($this->checkSsoClient()) {
                 $this->sso_mandatory = 1;
                 $username = $this->sso_username;
-                if (isset($this->options_sso['sso_username_pattern']) &&
-                    $this->options_sso['sso_username_pattern'] != ''
+                if (isset($this->options_sso['sso_username_pattern'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
                if (!empty($this->options_sso['sso_username_pattern'])) {
```

File: www/class/centreonAuth.SSO.class.php [hunk 6]
```diff
@@ -55,16 +55,18 @@ public function __construct(
...
+            && isset($_SERVER[$this->options_sso['sso_header_username']])
         ) {
             $this->sso_username = $_SERVER[$this->options_sso['sso_header_username']];
             if ($this->checkSsoClient()) {
                 $this->sso_mandatory = 1;
                 $username = $this->sso_username;
-                if (isset($this->options_sso['sso_username_pattern']) &&
-                    $this->options_sso['sso_username_pattern'] != ''
+                if (isset($this->options_sso['sso_username_pattern']) 
+                    && $this->options_sso['sso_username_pattern'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 7]
```diff
@@ -55,16 +55,18 @@ public function __construct(
...
                 ) {
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 8]
```diff
@@ -73,6 +75,53 @@ public function __construct(
                     );
                 }
             }
+        } elseif (isset($this->options_sso['keycloak_enable'])
+            && $this->options_sso['keycloak_enable'] == 1
+            && isset($this->options_sso['keycloak_url'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && !empty($this->options_sso['keycloak_url'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 9]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && $this->options_sso['keycloak_url'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 10]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && isset($this->options_sso['keycloak_redirect_url'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && !empty($this->options_sso['keycloak_redirect_url'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 11]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && $this->options_sso['keycloak_redirect_url'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 12]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && isset($this->options_sso['keycloak_realm'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && !empty($this->options_sso['keycloak_realm'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 13]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && $this->options_sso['keycloak_realm'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 14]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && isset($this->options_sso['keycloak_client_id'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && !empty($this->options_sso['keycloak_client_id'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 15]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && $this->options_sso['keycloak_client_id'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 16]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && isset($this->options_sso['keycloak_client_secret'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && !empty($this->options_sso['keycloak_client_secret'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 17]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            && $this->options_sso['keycloak_client_secret'] != ''
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 18]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+            $client_secret = $this->options_sso['keycloak_client_secret'];
+            $realm = $this->options_sso['keycloak_realm'];
+            $base = $this->options_sso['keycloak_url'];
+            $redirectNoEncode = $this->options_sso['keycloak_redirect_url'];
+
+            $redirect = urlencode($redirectNoEncode);
+            $authUrl = $base . "/realms/".$realm."/protocol/openid-connect/auth?client_id=" . $client_id
+                ."&response_type=code&redirect_uri=" . $redirect;
+
+            if (isset($_GET['force'])) {
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]: please sanitize this variable

File: www/class/centreonAuth.SSO.class.php [hunk 19]
```diff
@@ -73,6 +75,53 @@ public function __construct(
...
+                header('Location: ' . $authUrl);
+            }
+
+            if (isset($_GET['code'])) {
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]: please sanitize this variable

File: www/class/centreonAuth.SSO.class.php [hunk 20]
```diff
@@ -94,7 +143,9 @@ public function __construct(
 
     protected function checkSsoClient()
     {
-        if (isset($this->options_sso['sso_mode']) && $this->options_sso['sso_mode'] == 1) {
+        if (isset($this->options_sso['sso_enable']) && $this->options_sso['sso_enable'] == 1
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
        if (isset($this->options_sso['sso_enable'])
            && $this->options_sso['sso_enable'] == 1
```

File: www/class/centreonAuth.SSO.class.php [hunk 21]
```diff
@@ -94,7 +143,9 @@ public function __construct(
...
+            && isset($this->options_sso['sso_mode']) && $this->options_sso['sso_mode'] == 1
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && isset($this->options_sso['sso_mode'])
            && $this->options_sso['sso_mode'] == 1
```

File: www/class/centreonAuth.SSO.class.php [hunk 22]
```diff
@@ -106,6 +157,32 @@ protected function checkSsoClient()
             }
 
             $whitelist = explode(',', $this->options_sso['sso_trusted_clients']);
+            if(empty($whitelist[0])){
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            if (empty($whitelist[0])) {
```

File: www/class/centreonAuth.SSO.class.php [hunk 23]
```diff
@@ -106,6 +157,32 @@ protected function checkSsoClient()
...
+                return 1;
+            }
+            foreach ($whitelist as $value) {
+                $value = trim($value);
+                if ($value != "" && preg_match('/' . $value . '/', $_SERVER['REMOTE_ADDR'])) {
+                    return 1;
+                }
+            }
+        } elseif (isset($this->options_sso['keycloak_enable']) && $this->options_sso['keycloak_enable'] == 1
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
        } elseif (isset($this->options_sso['keycloak_enable'])
            && $this->options_sso['keycloak_enable'] == 1
```

File: www/class/centreonAuth.SSO.class.php [hunk 24]
```diff
@@ -106,6 +157,32 @@ protected function checkSsoClient()
...
+            && isset($this->options_sso['keycloak_mode']) && $this->options_sso['keycloak_mode'] == 1
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            && isset($this->options_sso['keycloak_mode'])
            && $this->options_sso['keycloak_mode'] == 1
```

File: www/class/centreonAuth.SSO.class.php [hunk 25]
```diff
@@ -106,6 +157,32 @@ protected function checkSsoClient()
...
+        ) {
+            # Mixed
+
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
```

File: www/class/centreonAuth.SSO.class.php [hunk 26]
```diff
@@ -106,6 +157,32 @@ protected function checkSsoClient()
...
+            foreach ($whitelist as $value) {
+                $value = trim($value);
+                if ($value != "" && preg_match('/' . $value . '/', $_SERVER['REMOTE_ADDR'])) {
+                    return 1;
+                }
+            }
+        } elseif (isset($this->options_sso['keycloak_enable']) && $this->options_sso['keycloak_enable'] == 1
+            && isset($this->options_sso['keycloak_mode']) && $this->options_sso['keycloak_mode'] == 1
+        ) {
+            # Mixed
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]: what does this mean ?

File: www/class/centreonAuth.SSO.class.php [hunk 27]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
             return parent::checkPassword($password, $token);
         }
     }
+
+
+    /**
+     * Connect to Keycloak and get token access
+     *
+     * @parem string $base Keycloak Server Url
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
     * @param string $base Keycloak Server Url
```

File: www/class/centreonAuth.SSO.class.php [hunk 28]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+     * @parem string $realm Keycloak Client Realm
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
     * @param string $realm Keycloak Client Realm
```

File: www/class/centreonAuth.SSO.class.php [hunk 29]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+     * @parem string $redirect_uri Keycloak Redirect Url
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]: @parem -> @ param ...

File: www/class/centreonAuth.SSO.class.php [hunk 30]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+     * @parem string $client_id Keycloak Client ID
+     * @parem string $client_secret Keycloak Client Secret
+     * @parem string $code Keycloak Authorization Code
+     *
+     * @return string
+    */
+    function getKeycloakToken($base, $realm, $redirect_uri, $client_id, $client_secret, $code) {
+
+        $url = $base."/realms/".$realm."/protocol/openid-connect/token";
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
        $url = $base . "/realms/" . $realm . "/protocol/openid-connect/token";
```

File: www/class/centreonAuth.SSO.class.php [hunk 31]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+        $data = array(
+            "client_id" => $client_id,
+            "client_secret" => $client_secret,
+            "grant_type" => "authorization_code",
+            "code" => $code,
+            "redirect_uri" => $redirect_uri);
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
            "redirect_uri" => $redirect_uri
        );
```

File: www/class/centreonAuth.SSO.class.php [hunk 32]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+
+        $resp = json_decode($result, true);
+
+        return $resp["access_token"];
+    }
+
+    /**
+     * Validate Centreon user on Keycloak
+     *
+     * @parem string $base Keycloak Server Url
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]: @parem -> @param ...

File: www/class/centreonAuth.SSO.class.php [hunk 33]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+     * @parem string $realm Keycloak Client Realm
+     * @parem string $client_id Keycloak Client ID
+     * @parem string $client_secret Keycloak Client Secret
+     * @parem string $token Keycloak Token Access
+     *
+     * @return string
+     */
+    function getKeycloakUserInfo($base, $realm, $client_id, $client_secret, $token){
+
+        $url = $base."/realms/".$realm."/protocol/openid-connect/token/introspect";
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
        $url = $base . "/realms/" . $realm . "/protocol/openid-connect/token/introspect";
```

File: www/class/centreonAuth.SSO.class.php [hunk 34]
```diff
@@ -135,4 +212,68 @@ protected function checkPassword($password, $token = "", $autoimport = false)
...
+        $data = array("token" => $token, "client_id" => $client_id, "client_secret" => $client_secret);
+
+        $ch = curl_init($url);
+        curl_setopt($ch, CURLOPT_HTTPHEADER,array("Authorization" => "Bearer ".$token));
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php]:
```suggestion
        curl_setopt($ch, CURLOPT_HTTPHEADER, array("Authorization" => "Bearer " . $token));
```

File: www/include/Administration/parameters/general/form.php [hunk 35]
```diff
@@ -214,6 +215,34 @@
...
+/*
+ * Keycloak
+ */
+$keycloak_enable[] = $form->createElement(
+    'checkbox',
+    'yes',
+    '&nbsp;',
+    '',
+    array("onchange" => "javascript:confirm("
+        ."'Are you sure you want to change this parameter ? Please read the help before.')"
```
Comment (sc979) [inline www/include/Administration/parameters/general/form.php]:
```suggestion
        . "'Are you sure you want to change this parameter ? Please read the help before.')"
```

Comment (sc979) [reply www/include/Administration/parameters/general/form.php]: you should use a gettext message to be able to translate the message.
ie : you can use `<?= _("yourText"); ?>`

File: www/include/core/login/login.php [hunk 36]
```diff
@@ -46,6 +46,15 @@
 $DBRESULT = $pearDB->query("SELECT `value` FROM `informations` WHERE `key` = 'version' LIMIT 1");
 $release = $DBRESULT->fetchRow();
 
+/**
+ * Getting Keycloak login state
+ */
+$DBRESULT = $pearDB->query("SELECT `value` FROM `options` WHERE `key` = 'keycloak_enable' LIMIT 1");
```
Comment (sc979) [inline www/include/core/login/login.php]: please use camelCase for your variables

File: www/include/core/login/login.php [hunk 37]
```diff
@@ -46,6 +46,15 @@
...
+$keycloakEnabled = $DBRESULT->fetchRow()["value"];
```
Comment (sc979) [inline www/include/core/login/login.php]: please use fetch instead of fetchRow

File: www/include/core/login/processLogin.php [hunk 38]
```diff
@@ -44,7 +44,8 @@
         $_POST["autologin"] &&
         isset($generalOptions["enable_autologin"]) &&
         $generalOptions["enable_autologin"])
-    || (!isset($generalOptions['sso_enable']) || $generalOptions['sso_enable'] == 1)) {
+    || (!isset($generalOptions['sso_enable']) || $generalOptions['sso_enable'] == 1)
```
Comment (sc979) [inline www/include/core/login/processLogin.php]: you need to choose if you want the boolean operators to be at the end or the beginning of a line

File: www/include/core/login/login.php [hunk 39]
```diff
@@ -46,6 +46,15 @@
...
+
+$DBRESULT = $pearDB->query("SELECT `value` FROM `options` WHERE `key` = 'keycloak_mode' LIMIT 1");
+$keycloakMode = $DBRESULT->fetchRow()["value"];
```
Comment (sc979) [inline www/include/core/login/login.php]: what is $keycloakMode for ? -> login.ihtml +26
is it used for the mixed authentication ?

**Comment by sc979** [comment]
Please add the french translation and the improve the translation template of this message :
`Are you sure you want to change this parameter ? Please read the help before.`

Please add the french translations for the help tooltips and the translation template for the other languages

**Comment by sc979** [comment]
You should add the other languages translation template too

**Comment by lgcosta** [comment]
> You should add the other languages translation template too

add the variables for pt_BR and pt_PT that I translate into review ;-)

**Comment by lpinsivy** [comment]
Can you translate this in ES, and PTs?:

```
msgid "Are you sure you want to change this parameter? Please read the help before."
msgstr "Êtes-vous sûr de vouloir changer ce paramètre ? Veuillez lire l'aide avant."

 msgid "Keycloak trusted client addresses"
msgstr "Adresses des clients de confiance SSO Keycloak"

 msgid "Keycloak blacklist client addresses"
msgstr "Adresses des clients de bloqués Keycloak"

 msgid "Keycloak Server Url"
msgstr "URL du serveur Keycloak"

 msgid "Keycloak Redirect Url"
msgstr "Adresse de redirection vers Centreon depuis Keycloak"

 msgid "Keycloak Client Realm"
msgstr "Royaume client Keycloak"

 msgid "Keycloak Client ID"
msgstr "Identifiant client Keycloak"

 msgid "Keycloak Client Secret"
msgstr "Secret client Keycloak"

```

**Comment by sc979** [comment]
Translation templates added for messages and help tooltip

**Comment by lpinsivy** [comment]
ES, PT et PT_BR added from @lgcosta branch

**Comment by sc979** [comment]
conflits resolved and branch rebased on August 5th

File: www/include/core/login/login.php [hunk 40]
```diff
@@ -43,8 +43,17 @@
...
  */
-$DBRESULT = $pearDB->query("SELECT `value` FROM `informations` WHERE `key` = 'version' LIMIT 1");
-$release = $DBRESULT->fetch();
+$result = $pearDB->query("SELECT `value` FROM `informations` WHERE `key` = 'version' LIMIT 1");
+$release = $result->fetch();
+
+/**
+ * Getting Keycloak login state
+ */
+$result = $pearDB->query("SELECT `value` FROM `options` WHERE `key` = 'keycloak_enable' LIMIT 1");
```
Comment (sc979) [inline www/include/core/login/login.php]: couldn't you merge these two queries using IN instead of equal ?

**Pull request merged**